src/dbnode/encoding/proto/custom_unmarshal.go

```diff
@@ -0,0 +1,379 @@
+			return fmt.Errorf("encountered unknown field with field number: %d", fieldNum)
+		}
+
+		isCustomField := u.isCustomField(fd)
+		if !isCustomField {
+			if u.nonCustomValues == nil {
+				u.nonCustomValues = dynamic.NewMessage(u.schema)
+			}
+
+			_, err = u.skip(wireType)
+			if err != nil {
+				return err
+			}
+
+			var (
+				startIdx       = tagAndWireTypeStartOffset
+				endIdx         = u.decodeBuf.index
+				marshaledField = u.decodeBuf.buf[startIdx:endIdx]
+			)
+			// A marshaled Protobuf message consists of a stream of <fieldNumber, wireType, value>
+			// tuples, all of which are optional, with no additional header or footer information.
+			// This means that each tuple within the stream can be thought of as its own complete
+			// marshaled message and as a result we can build up the nonCustomValues *dynamic.Message
+			// one field at a time by calling UnmarshalMerge() on sub-slices that contain a complete
+			// tuple.
+			u.nonCustomValues.UnmarshalMerge(marshaledField)
+			u.numNonCustom++
+			continue
+		}
+
+		value, err := u.unmarshalCustomField(fd, wireType)
+		if err != nil {
+			return err
+		}
+
+		if isSorted && len(u.customValues) > 1 {
+			// Check if the slice is sorted as it's built to avoid resorting
+			// unnecessarily at the end.
+			lastFieldNum := u.customValues[len(u.customValues)-1].fieldNumber
+			if fieldNum < lastFieldNum {
+				isSorted = false
+			}
+		}
+
+		u.customValues = append(u.customValues, value)
+	}
+
+	u.decodeBuf.reset(u.decodeBuf.buf)
+
+	if !isSorted {
+		// Avoid resorting if possible.
+		sort.Sort(u.customValues)
+	}
+
+	return nil
+}
+
+func (u *customUnmarshaler) isCustomField(fd *desc.FieldDescriptor) bool {
+	if fd.IsRepeated() || fd.IsMap() {
+		// Map should always be repeated but include the guard just in case.
+		return false
+	}
+
+	if fd.GetMessageType() != nil {
+		// Skip nested messages.
+		return false
+	}
+
+	return true
+}
+
+// skip will skip over the next value in the encoded stream (given that the tag and
+// wiretype have already been decoded).
+func (u *customUnmarshaler) skip(wireType int8) (int, error) {
+	switch wireType {
+	case proto.WireFixed32:
+		numSkipped := 4
+		u.decodeBuf.index += numSkipped
+		return numSkipped, nil
+
+	case proto.WireFixed64:
+		numSkipped := 8
+		u.decodeBuf.index += numSkipped
+		return numSkipped, nil
+
+	case proto.WireVarint:
+		var (
+			numSkipped               = 0
+			offsetBeforeDecodeVarInt = u.decodeBuf.index
+		)
+		_, err := u.decodeBuf.decodeVarint()
+		if err != nil {
+			return numSkipped, err
+		}
+		numSkipped += u.decodeBuf.index - offsetBeforeDecodeVarInt
+		return numSkipped, nil
+
+	case proto.WireBytes:
+		var (
+			numSkipped                 = 0
+			offsetBeforeDecodeRawBytes = u.decodeBuf.index
+		)
+		// Bytes aren't copied because they're just being skipped over so
+		// copying would be wasteful.
+		_, err := u.decodeBuf.decodeRawBytes(false)
+		if err != nil {
+			return numSkipped, err
```

Comment:
nit: `0`, just to clearly communicate that the value (even though we can see that the var is set to 0) is invalid